examples/ex18.cpp

```diff
@@ -250,60 +253,259 @@ int main(int argc, char *argv[])
       // maximum char speed at all quadrature points on all faces.
-      Vector z(A.Width());
-      max_char_speed = 0.;
-      A.Mult(sol, z);
-      dt = cfl * hmin / max_char_speed / (2*order+1);
+      Vector z(sol.Size());
+      euler.Mult(sol, z);
+
+      double max_char_speed = euler.GetMaxCharSpeed();
+      dt = cfl * hmin / max_char_speed / (2 * order + 1);
    }
 
    // Integrate in time.
    bool done = false;
-   for (int ti = 0; !done; )
+   for (int ti = 0; !done;)
    {
       double dt_real = min(dt, t_final - t);
 
       ode_solver->Step(sol, t, dt_real);
       if (cfl > 0)
       {
-         dt = cfl * hmin / max_char_speed / (2*order+1);
+         double max_char_speed = euler.GetMaxCharSpeed();
+         dt = cfl * hmin / max_char_speed / (2 * order + 1);
       }
       ti++;
 
-      done = (t >= t_final - 1e-8*dt);
+      done = (t >= t_final - 1e-8 * dt);
       if (done || ti % vis_steps == 0)
       {
-         cout << "time step: " << ti << ", time: " << t << endl;
+         out << "time step: " << ti << ", time: " << t << endl;
          if (visualization)
          {
+            GridFunction mom(&dfes, sol.GetData());
             sout << "solution\n" << mesh << mom << flush;
+            sout << "window_title 't = " << t << "'";
          }
       }
    }
 
    tic_toc.Stop();
-   cout << " done, " << tic_toc.RealTime() << "s." << endl;
+   out << " done, " << tic_toc.RealTime() << "s." << endl;
 
    // 9. Save the final solution. This output can be viewed later using GLVis:
-   //    "glvis -m vortex.mesh -g vortex-1-final.gf".
-   for (int k = 0; k < num_equation; k++)
+   //    "glvis -m euler.mesh -g euler-1-final.gf".
    {
-      GridFunction uk(&fes, u_block.GetBlock(k));
-      ostringstream sol_name;
-      sol_name << "vortex-" << k << "-final.gf";
-      ofstream sol_ofs(sol_name.str().c_str());
-      sol_ofs.precision(precision);
-      sol_ofs << uk;
+      ostringstream mesh_name;
+      mesh_name << "euler-mesh-final.mesh";
+      ofstream mesh_ofs(mesh_name.str().c_str());
+      mesh_ofs.precision(precision);
+      mesh_ofs << mesh;
+
+      for (int k = 0; k < num_equations; k++)
+      {
+         GridFunction uk(&fes, sol.GetData() + k * fes.GetNDofs());
+         ostringstream sol_name;
+         sol_name << "euler-" << k << "-final.gf";
+         ofstream sol_ofs(sol_name.str().c_str());
+         sol_ofs.precision(precision);
+         sol_ofs << uk;
+      }
    }
 
    // 10. Compute the L2 solution error summed for all components.
-   if (t_final == 2.0)
-   {
-      const double error = sol.ComputeLpError(2, u0);
-      cout << "Solution error: " << error << endl;
-   }
+   //   if (t_final == 2.0) {
+   const double error = sol.ComputeLpError(2, u0);
+   out << "Solution error: " << error << endl;
 
    // Free the used memory.
    delete ode_solver;
+   delete numericalFlux;
 
    return 0;
 }
+
+void EulerMesh(const int problem, const char **mesh_file)
```

Comment:
As mentioned above, I think it would be better if this was:

```suggestion
Mesh EulerMesh(const int problem, const string &mesh_file)
```